### PR TITLE
Make testimony tokens configurable

### LINKED ISSUE #100: Custom required docstring tags
Make testimony configurable so one can define what are the minimum valid docstring. Also one may want to just have `@test` as the minimum valid docstring as well as the basic ones and plus some other custom tags.

### PR DESCRIPTION
**Pull request by elyezer**
Improve testimony to accept custom tokens and minimum required tokens for each
docstrings.

Close #100

### CODE REVIEW

**Comment by elyezer** [comment]
This is showing to be very hard to implement since there are some testimony reports that expects some tags like the status to be manual or automated in order to make a test method automated or not automated.

If we really decide to make testimony configurable, we will need to improve the reports and make them configurable or maybe drop some of them.

I will continue working on this and check what should be updated and what should be dropped.

**Comment by elyezer** [comment]
I wanna take another review before removing the WIP, but the PR is ready to be reviewed.

File: testimony/__init__.py [hunk 1]
```diff
@@ -63,26 +57,17 @@ def prefixed_lines():
 
 
 def is_test_module(filename):
-    """Return ``True`` if the ``filename`` matches an expected test
-    module name.
-    """
+    """Indicate if ``filename`` match a test module file name."""
```
Comment (sthirugn) [inline testimony/__init__.py]: How about:
```
"""Indicates if the given ``filename`` is a test module."""
```

File: testimony/__init__.py [hunk 2]
```diff
@@ -63,26 +57,17 @@ def prefixed_lines():
...
 class TestFunction(object):
-    """Class that wraps a ``ast.FunctionDef`` instance and provide useful
-    information for creating the reports.
-
-    An instance of ``TestFunction`` will proxy attribute lookups to the wrapped
-    ``ast.FunctionDef`` in order to make easy access to atribute like
-    ``name``.
+    """Wrapper for ``ast.FunctionDef`` which parse docstring information.
 
+    The information on ``tokens`` and ``invalid_tokens`` can be used to provide
```
Comment (sthirugn) [inline testimony/__init__.py]: How about:
```
The ``tokens`` and ``invalid_tokens`` parameters provide meaningful information about the test.
```

File: testimony/__init__.py [hunk 3]
```diff
@@ -92,190 +77,66 @@ def __init__(self, function_def, parent_class=None, testmodule=None):
...
-            if tag == 'assert':
-                tag = 'assertion'
-            if tag == 'type':
-                tag = 'test_type'
-            setattr(self, tag, value)
-        self.unexpected_tags = unexpected_tags
+        # Parse module, class and function docstrings. Every loop update the
+        # already defined tokens and invalid_tokens. Doing in the mentioned
+        # order will ensure that function docstrings have more priority over
+        # class and module docstring respectively.
```
Comment (sthirugn) [inline testimony/__init__.py]: Parse module, class and function docstrings.  Every loop updates the already defined tokens and invalid_tokens. The order of processing ensures that function docstring has more priority over class and module docstrings respectively.

File: testimony/__init__.py [hunk 4]
```diff
@@ -92,190 +77,66 @@ def __init__(self, function_def, parent_class=None, testmodule=None):
...
     @property
     def has_valid_docstring(self):
-        """Returns ``True`` if at least feature and assert tags are defined.
-        """
-        return (
-            self.assertion is not None and
-            self.feature is not None and
-            self.test is not None
-        )
+        """Boolean indicating if docstring have the minimum tokens."""
```
Comment (sthirugn) [inline testimony/__init__.py]: To be consistent with other docstring:
Indicates if the docstring has the minimum tokens.

File: testimony/__init__.py [hunk 5]
```diff
@@ -92,190 +77,66 @@ def __init__(self, function_def, parent_class=None, testmodule=None):
...
-            'skipped-lines': self.skipped_lines,
-            'status': self.status,
-            'steps': self.steps,
-            'tags': self.tags,
-            'test': self.test,
-            'test_type': self.test_type,
-            'unexpected-tags': self.unexpected_tags,
-        }
+        """Return tokens dictionary and include invalid-tokens as an item."""
+        values = self.tokens.copy()
```
Comment (sthirugn) [inline testimony/__init__.py]: why not values['tokens'] here to be consistent with next line?
```
return_value = {}
return_value['tokens'] = self.tokens.copy()
return_value['invalid-tokens'] = self.invalid_tokens.copy()
```

File: testimony/parser.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            raise ValueError('tokens should contain minimum_tokens')
+
+    def parse(self, docstring=None):
+        """Parses the docstring and returns the valid and invalid tokens.
+
+        For example in the following docstring (using single quote to demo)::
+
+            '''First line
+
+            @valid_tag1: value
```
Comment (sthirugn) [inline testimony/parser.py]: s/value/value1 ?

File: testimony/parser.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            @valid_tag2: value2
+            @invalid_tag1: value
```
Comment (sthirugn) [inline testimony/parser.py]: s/value/value1 ?

**Comment by sthirugn** [comment]
Great work!

Pending comments.

File: testimony/__init__.py (hunk 1 above)
Comment (elyezer) [inline testimony/__init__.py]: A good practice is to use on imperative mode. I'm following it.

> The docstring is a phrase ending in a period. It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...".

From https://www.python.org/dev/peps/pep-0257/

File: testimony/__init__.py (hunk 4 above)
Comment (elyezer) [inline testimony/__init__.py]: I will update, but will use `Indicate if the docstring has the minimum tokens`

File: testimony/__init__.py (hunk 5 above)
Comment (elyezer) [inline testimony/__init__.py]: You have a point, initially I though on having the tokens keys directly on the return dict, but it seems to make more sense to have tokens and invalid-tokens keys.

**Comment by elyezer** [comment]
Thank you @sthirugn for all the comments. I will update them PR.

**Comment by sthirugn** [comment]
ACK pending Issue #104

File: CHANGELOG [hunk 8]
```diff
@@ -0,0 +1,35 @@
+v1.0.3
```
Comment (elyezer) [inline CHANGELOG]: Adding a CHANGELOG file to track changes, we may raw include it later on in documentation if needed.

File: docs/index.rst [hunk 9]
```diff
@@ -1,380 +1,150 @@
...
-  	optional arguments:
-  	-h, --help            show this help message and exit
-  	-j, --json            JSON output
-  	-n, --nocolor         Do not use color option
-	-t [TAGS [TAGS ...]], --tags [TAGS [TAGS ...]]
-                          space separated tags to search.  Note: Always run this
-                          only in the root of the project where test cases are
-                          stored
-	
-print - test cases::
```
Comment (elyezer) [inline docs/index.rst]: Removed the outputs from the documentation since it may change and the documentation do not match the current output.

File: testimony/__init__.py [hunk 10]
```diff
@@ -1,6 +1,6 @@
 # -*- encoding: utf-8 -*-
 """DocString manipulation methods to create test reports"""
-from __future__ import print_function
+from __future__ import print_function, unicode_literals
```
Comment (elyezer) [inline testimony/__init__.py]: Let's go one step closer to Python 3.

**Comment by omaciel** [comment]
README needs a bit of TLC but otherwise, ACK

**Comment by sthirugn** [comment]
I agree README needs polishing but not a blocker

**Comment by sthirugn** [comment]
ACK
